File: assignment7/program.py
class StatisticsCalculator:
    """
    A class to calculate basic statistics for a list of integers.
    Demonstrates object-oriented programming principles.
    """
    
    def __init__(self, data=None):
        """
        Initialize the calculator with optional data.
        
        Args:
            data (list): Optional list of integers
        """
        self.data = data if data is not None else []
    
    def set_data(self, data):
        """
        Set or update the data for calculations.
        
        Args:
            data (list): List of integers
        """
        if not isinstance(data, list):
            raise TypeError("Data must be a list")
        self.data = data.copy()  # Create a copy to prevent external modifications
    
    def get_data(self):
        """
        Return a copy of the current data.
        
        Returns:
            list: Copy of the current data
        """
        return self.data.copy()
    
    def calculate_mean(self):
        """
        Calculate the mean (average) of the data.
        
        Returns:
            float: Mean value, or 0 if data is empty
        """
        if not self.data:
            return 0
        return sum(self.data) / len(self.data)
    
    def calculate_median(self):
        """
        Calculate the median (middle value) of the data.
        
        Returns:
            float: Median value, or 0 if data is empty
        """
        if not self.data:
            return 0
        
        # Create a sorted copy of the data
        sorted_data = sorted(self.data)
        n = len(sorted_data)
        
        # Handle both even and odd lengths
        if n % 2 == 0:
            # Even length - average the middle two values
            return (sorted_data[n//2 - 1] + sorted_data[n//2]) / 2
        else:
            # Odd length - return the middle value
            return sorted_data[n//2]
    
    def calculate_mode(self):
        """
        Calculate the mode (most frequent value(s)) of the data.
        
        Returns:
            tuple: (list of mode values, count of occurrences)
                   Returns ([], 0) if there is no mode (all values appear once)
        """
        if not self.data:
            return [], 0
            
        # Count occurrences of each value
        counts = {}
        for value in self.data:
            if value in counts:
                counts[value] += 1
            else:
                counts[value] = 1
                
        # Find the maximum frequency
        max_count = max(counts.values())
        
        # If all values appear only once, there is no mode
        if max_count == 1:
            return [], 0
            
        # Get all values that appear with the maximum frequency
        modes = [key for key, value in counts.items() if value == max_count]
        
        return modes, max_count
    
    def print_statistics(self):
        """
        Calculate and print all statistics.
        """
        mean = self.calculate_mean()
        median = self.calculate_median()
        modes, mode_count = self.calculate_mode()
        
        print("\nStatistics:")
        print(f"Mean: {mean:.2f}")
        print(f"Median: {median:.2f}")
        
        print("Mode: ", end="")
        if mode_count == 0:
            print("No mode (all values appear once)")
        else:
            print(f"{modes} (appears {mode_count} times)")


# Demonstration of the StatisticsCalculator class
if __name__ == "__main__":
    # Create an instance of the calculator
    calculator = StatisticsCalculator()
    
    # Get input from the user
    try:
        user_input = input("Enter integers separated by spaces: ")
        data = [int(x) for x in user_input.split()]
        
        # Set the data and calculate statistics
        calculator.set_data(data)
        calculator.print_statistics()
        
    except ValueError:
        print("Error: Please enter valid integers separated by spaces.")

File: assignment7/program.c
#include <stdio.h>
#include <stdlib.h>

// Function to compare integers for qsort
int compare(const void *a, const void *b) {
    return (*(int *)a - *(int *)b);
}

// Calculate mean of an array of integers
double calculate_mean(int *array, int size) {
    if (size == 0) return 0;
    
    long sum = 0;
    for (int i = 0; i < size; i++) {
        sum += array[i];
    }
    
    return (double)sum / size;
}

// Calculate median of an array of integers
double calculate_median(int *array, int size) {
    if (size == 0) return 0;
    
    // Sort the array
    qsort(array, size, sizeof(int), compare);
    
    // Return the median
    if (size % 2 == 0) {
        // Even number of elements, average the middle two
        return (array[size/2 - 1] + array[size/2]) / 2.0;
    } else {
        // Odd number of elements, return the middle one
        return array[size/2];
    }
}

// Calculate mode of an array of integers
void calculate_mode(int *array, int size, int *mode_values, int *mode_count, int *unique_modes) {
    if (size == 0) {
        *unique_modes = 0;
        return;
    }
    
    // Sort the array to group identical values
    qsort(array, size, sizeof(int), compare);
    
    int max_frequency = 0;
    int current_frequency = 1;
    *unique_modes = 0;
    
    // Find the maximum frequency
    for (int i = 1; i <= size; i++) {
        if (i < size && array[i] == array[i-1]) {
            current_frequency++;
        } else {
            if (current_frequency > max_frequency) {
                max_frequency = current_frequency;
                *unique_modes = 1;
                mode_values[0] = array[i-1];
            } else if (current_frequency == max_frequency) {
                mode_values[*unique_modes] = array[i-1];
                (*unique_modes)++;
            }
            current_frequency = 1;
        }
    }
    
    *mode_count = max_frequency;
}

int main() {
    int size;
    printf("Enter the number of integers: ");
    scanf("%d", &size);
    
    if (size <= 0) {
        printf("Invalid size. Please enter a positive number.\n");
        return 1;
    }
    
    // Dynamically allocate memory for the array
    int *array = (int *)malloc(size * sizeof(int));
    if (array == NULL) {
        printf("Memory allocation failed.\n");
        return 1;
    }
    
    printf("Enter %d integers:\n", size);
    for (int i = 0; i < size; i++) {
        scanf("%d", &array[i]);
    }
    
    // Calculate statistics
    double mean = calculate_mean(array, size);
    double median = calculate_median(array, size);
    
    // For mode calculation
    int *mode_values = (int *)malloc(size * sizeof(int)); // Max possible unique values
    int mode_count = 0;
    int unique_modes = 0;
    
    calculate_mode(array, size, mode_values, &mode_count, &unique_modes);
    
    // Display results
    printf("\nStatistics:\n");
    printf("Mean: %.2f\n", mean);
    printf("Median: %.2f\n", median);
    
    printf("Mode: ");
    if (mode_count == 1) {
        printf("No mode (all values appear once)\n");
    } else {
        for (int i = 0; i < unique_modes; i++) {
            printf("%d ", mode_values[i]);
        }
        printf("(appears %d times)\n", mode_count);
    }
    
    // Free allocated memory
    free(array);
    free(mode_values);
    
    return 0;
}